### ISSUE TITLE
Running rustdoc on code using quote generates false lint warnings for unused_braces

### ISSUE THREAD

**Issue by shepmaster** [rust-lang/rust#70814]
I tried this code:

```toml
[package]
name = "repro"
version = "0.1.0"
edition = "2018"

[dependencies]
quote = "=1.0.3"
```

```rust
use quote::quote;

pub fn repro() {
    let many = [1, 2, 3];

    let _together = quote! {
        #(#many),*
    };
}
```

I think this is a basic and common usage of the [quote](https://crates.io/crates/quote) crate.

I expected to see no warnings:

```
% RUSTFLAGS=-Dwarnings cargo +nightly-2020-04-04 build
    Finished dev [unoptimized + debuginfo] target(s) in 0.01s

% RUSTFLAGS=-Dwarnings cargo +nightly-2020-04-04 doc
 Documenting repro v0.1.0 (/private/tmp/repro)
    Finished dev [unoptimized + debuginfo] target(s) in 0.82s
```

Instead, warnings are generated (and my CI build fails due to the `deny`):

```
% RUSTFLAGS=-Dwarnings cargo +nightly-2020-04-04 build
    Finished dev [unoptimized + debuginfo] target(s) in 0.01s

% RUSTFLAGS=-Dwarnings cargo +nightly-2020-04-04 doc
 Documenting repro v0.1.0 (/private/tmp/repro)
warning: unnecessary braces around block return value
  |
  = note: `#[warn(unused_braces)]` on by default

    Finished dev [unoptimized + debuginfo] target(s) in 0.82s
```

Note that the warning only appears when running rustdoc.

### Meta

`rustc +nightly-2020-04-04 --version --verbose`:

```
rustc 1.44.0-nightly (74bd074ee 2020-04-03)
binary: rustc
commit-hash: 74bd074eefcf4915c73d1ab91bc90859664729e6
commit-date: 2020-04-03
host: x86_64-apple-darwin
release: 1.44.0-nightly
LLVM version: 9.0
```

**Comment by eddyb** [comment]
> Note that the warning only appears when running rustdoc.

Hang on, could this be because of the transformation `rustdoc` uses (AFAIK) that replaces `fn` bodies with `loop {}`?

**Comment by Nemo157** [comment]
Simplified repro source, both inner blocks are necessary for some reason:

```rust
pub fn repro() {
    {
        {
            use std;
        }
    }
}
```

**Comment by eddyb** [comment]
@Nemo157 My guess it's because `{{}}` is basically `{()}` - the outer block has the inner block as its final expression. Curious that this is `rustdoc`-specific.

**Comment by tmandry** [comment]
I just hit a similar error message in rustc (not rustdoc). Unused braces lint with no span or other information. I don't have a reproducer currently, since it's from a big crate and I don't have much to go by.

**Comment by grantperry** [comment]
I'm getting `unnescessary braces around block return value` from check/clippy
```
warning: unnecessary braces around block return value
   --> src/models/user.rs:137:63
    |
137 |     fn ticket(&self, context : &SharedContext) -> Option<Ticket> { self.get_ticket(context) }
    |                                                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: remove these braces
```
Is this related or should I create a new issue?

Using Nightly 1.44.0 2020-04-16

**Comment by lcnr** [comment]
@grantperry is this inside a macro?

**Comment by grantperry** [comment]
@lcnr yes it is. Sounds like it's something you already know about?

**Comment by lcnr** [comment]
If you are using a proc macro using `quote` then this should be the same issue afaict

**Comment by lcnr** [comment]
I tried to fix this by checking if the surrounding expression comes from an expansion, but this didn't actually fix this. :thinking: @matklad do you have an idea/can this be fixed in quote?

**Comment by matklad** [comment]
I don't think I am familiar with the code in question @lcnr

**Comment by lcnr** [comment]
Sorry, mb :sweat: meant @dtolnay as he is the owner of `quote`.

**Comment by dtolnay** [comment]
Labeling <kbd>regression-from-stable-to-nightly</kbd>. I would consider accepting a workaround in quote if somebody who is not me sends a PR and it is a very simple fix, but really this needs to be fixed in rustdoc / rustc.

**Comment by LeSeulArtichaut** [comment]
Let’s try to find the cause of the regression
@rustbot ping cleanup

**Comment by AminArria** [comment]
searched nightlies: from nightly-2020-03-29 to nightly-2020-04-05
regressed nightly: nightly-2020-04-02
searched commits: from https://github.com/rust-lang/rust/commit/a5b09d35473615e7142f5570f5c5fad0caf68bd2 to https://github.com/rust-lang/rust/commit/76b11980ad416c3ad6143504c2277757ecacf9b5
regressed commit: https://github.com/rust-lang/rust/commit/58dd1ce8383aaebcad9b6027b89a316fd868b35c

I guess the PR would be #70081

**Comment by shepmaster** [comment]
I'm sorry, that was my bad for not including that information in the original comment. We already knew the cause of the regression (found in #70717), and the author of the regression is already aware (@lcnr) of the issue, but not yet a fix.

**Comment by spastorino** [comment]
Assigning `P-medium` as [discussed as part of the Prioritization Working Group process](https://rust-lang.zulipchat.com/#narrow/stream/227806-t-compiler.2Fwg-prioritization/topic/I-prioritize.20.2371202.20ICE.20when.20using.20const.20evaluation) and removing `I-prioritize`.

**Comment by osa1** [comment]
I think e42337b6089d a fixed this issue. It didn't add a test though so perhaps mark as needs-test?

**Comment by Igosuki** [comment]
@osa1 it didn't : 
```
warning: unnecessary braces around block return value
  --> strategies/src/types.rs:67:34
   |
67 |     fn asset_type(&self) -> &str { self.asset_type.as_ref() }
   |                                  ^^                        ^^
   |
help: remove these braces
   |
67 -     fn asset_type(&self) -> &str { self.asset_type.as_ref() }
67 +     fn asset_type(&self) -> &str self.asset_type.as_ref()
```
